python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -21,59 +27,208 @@ def __init__(
+        # Cached actor class.
+        self._cls = None
+        # Whether no more submittable bundles will be added.
+        self._inputs_done = False
 
     def start(self, options: ExecutionOptions):
         super().start(options)
+
         # Create the actor workers and add them to the pool.
-        ray_remote_args = self._apply_default_remote_args(self._ray_remote_args)
-        cls_ = ray.remote(**ray_remote_args)(MapWorker)
-        for _ in range(self._pool_size):
-            self._actor_pool.add_actor(cls_.remote())
+        self._cls = ray.remote(**self._ray_remote_args)(_MapWorker)
+        for _ in range(self._autoscaling_policy.min_workers):
+            self._start_actor()
 
-    def submit(
-        self, input_blocks: List[ObjectRef[Block]]
-    ) -> ObjectRef[ObjectRefGenerator]:
-        # Pick an actor from the pool.
-        actor = self._actor_pool.pick_actor()
-        # Submit the map task.
-        ref = actor.submit.options(num_returns="dynamic").remote(
-            self._transform_fn_ref, *input_blocks
-        )
-        self._active_actors[ref] = actor
-        return ref
+    def _start_actor(self):
+        """Start a new actor and add it to the actor pool as a pending actor."""
+        assert self._cls is not None
+        actor = self._cls.remote()
+        self._actor_pool.add_pending_actor(actor, actor.ready.remote())
+
+    def add_submittable_bundle(self, bundle: RefBundle):
+        self._bundle_queue.append(bundle)
+        # Try to dispatch all bundles in the queue, including this new bundle.
+        self._dispatch_tasks()
+
+    def _dispatch_tasks(self):
+        """Try to dispatch tasks from the bundle buffer to the actor pool.
+
+        This is called when:
+            * a new input bundle is added,
+            * a task finishes,
+            * a new worker has been created.
+        """
+        while self._bundle_queue:
+            # Pick an actor from the pool.
+            actor = self._actor_pool.pick_actor()
+            if actor is None:
+                # No actors available for executing the next task.
+                break
+            # Submit the map task.
+            bundle = self._bundle_queue.popleft()
+            input_blocks = [block for block, _ in bundle.blocks]
+            ref = actor.submit.options(num_returns="dynamic").remote(
+                self._transform_fn_ref, *input_blocks
+            )
+            task = TaskState(bundle)
+            self._tasks[ref] = (task, actor)
+            # Notify output queue that this task is pending.
+            self._output_queue.notify_pending_task(task)
+            # Update stats.
+            self._metrics.cur += bundle.size_bytes()
+            if self._metrics.cur > self._metrics.peak:
+                self._metrics.peak = self._metrics.cur
+
+        if self._bundle_queue:
+            # Try to scale up if work remains in the work queue.
+            self._scale_up_if_needed()
```

Comment:
I'm thinking that we might want to always try to scale up, even if we fully consume the work queue, since we'd ideally want the actor pool capacity to stay ahead of the work queue by 1 or more actor so the actor creation latency is mostly hidden (pipelined with mapping over previous bundles).